### ISSUE TITLE
Multi index

### ISSUE THREAD

**Issue by floridoo** [hashicorp/go-memdb#15]
It would be nice to be able to create multiple index entries for one object (e.g. by returning multiple values from `FromObject()`).

Example:

``` go
type Object struct {
    tags []string
}
```

If you want to have an index on the tags, that's not currently possible.

Would you be interested in a pull request for this?

**Comment by armon** [comment]
@floridoo Sure! That seems like an interesting enhancement, I'd be open to reviewing a PR for it!

**Comment by floridoo** [comment]
There are 2 possible approaches:

**Variant A:**
Change the `Indexer` interface to return an array of values from `FromObject`:
```go
type Indexer interface {
	// replaces FromObject(raw interface{}) (bool, []byte, error)
	FromObject(raw interface{}) (bool, [][]byte, error)
	FromArgs(args ...interface{}) ([]byte, error)
}
```
**Advantages**: simple to implement, simple to use
**Disadvantage**: breaks backwards compatibility for custom indexers

**Variant B:**
Add a new `MultiIndexer` interface:
```go
type MultiIndexer interface {
	FromObjectMulti(raw interface{}) (bool, [][]byte, error)
}
```
**Advantage**: doesn't break compatibility
**Disadvantages**: needs more checks for every insert and delete operation, less intuitive (multi indexers need to implement both the `FromObject` and the `FromObjectMulti` methods, even though just the second one is used)

**TLDR**: Is it ok to break backward compatibility by changing the `Indexer` interface?

**Comment by armon** [comment]
@floridoo So what I don't like with A is that it breaks backward compatibility. What I don't like with B is that the slice of slice of bytes will require an additional allocation in the common case. Most of our indexers will return a single byte slice but they will need to wrap it again, which will be wasteful.

What if we instead return `(bool, interface{}, error)` where the value must be one of `[]byte` or `[][]byte` and the indexer gets to pick. That should be backwards compatible and avoids the extra allocation.

**Comment by floridoo** [comment]
Thanks @armon, that sounds better. But it still breaks backward compatibility (needs changing the method signature for all Indexer implementations).

I sent a pull request with an alternative, that both keeps the backward compatibility and doesn't need wrapping the single values. Tell me what you think of it.

**Comment by slackpad** [comment]
Fixed in https://github.com/hashicorp/go-memdb/pull/17.
